compiler/src/dmd/identifier.d

```diff
@@ -317,23 +318,25 @@ nothrow:
      */
     extern (D) static bool isValidIdentifier(const(char)[] str)
     {
-        if (str.length == 0 ||
-            (str[0] >= '0' && str[0] <= '9')) // beware of isdigit() on signed chars
-        {
+        if (str.length == 0)
             return false;
+
+        size_t idx;
+        dchar dc;
+
+        {
+            const errorMessage = utf_decodeChar(str, idx, dc);
+            if (errorMessage || !isUAX31Start(dc))
+                return false;
         }
 
-        size_t idx = 0;
         while (idx < str.length)
         {
-            dchar dc;
-            const s = utf_decodeChar(str, idx, dc);
-            if (s ||
-                !((dc >= 0x80 && isUniAlpha(dc)) || isalnum(dc) || dc == '_'))
-            {
+            const errorMessage = utf_decodeChar(str, idx, dc);
+            if (errorMessage || !isUAX31Continue(dc))
```

Comment:
having the ASCII check preceding the isUAX... means that plain ASCII text will be fast.